src/target/xtensa/xtensa_chip.c

```diff
@@ -0,0 +1,252 @@
+	if ((dm_cfg->queue_tdi_idle == NULL) && (dm_cfg->tap != NULL)) {
+		dm_cfg->queue_tdi_idle = xtensa_chip_queue_tdi_idle;
+		dm_cfg->queue_tdi_idle_arg = target;
+	}
+	int ret = xtensa_init_arch_info(target, &xtensa_chip->xtensa, dm_cfg);
+	if (ret != ERROR_OK) {
+		return ret;
+	}
+	/* All xtensa target structures point back to original xtensa_chip */
+	xtensa_chip->xtensa.xtensa_chip = arch_info;
+	return ERROR_OK;
+}
+
+static int xtensa_chip_handle_target_event(struct target *target, enum target_event event, void *priv)
+{
+	if (target != priv)
+		return ERROR_OK;
+
+	LOG_DEBUG("%d", event);
+
+	int ret = xtensa_handle_target_event(target, event, priv);
+	if (ret != ERROR_OK)
+		return ret;
+
+	switch (event) {
+	case TARGET_EVENT_HALTED:
+		/* debug stubs not supported */
+		break;
+	case TARGET_EVENT_GDB_DETACH:
+		/* Flash BP not supported */
+		break;
+	default:
+		break;
+	}
+	return ERROR_OK;
+}
+
+int xtensa_chip_target_init(struct command_context *cmd_ctx, struct target *target)
+{
+	int ret = xtensa_target_init(cmd_ctx, target);
+	if (ret != ERROR_OK)
+		return ret;
+
+	ret = target_register_event_callback(xtensa_chip_handle_target_event, target);
+	if (ret != ERROR_OK)
+		return ret;
+
+	ret = xtensa_fileio_init(target);
+	if (ret != ERROR_OK)
+		return ret;
+
+	return ERROR_OK;
+}
+
+int xtensa_chip_arch_state(struct target *target)
+{
+	return ERROR_OK;
+}
+
+static int xtensa_chip_poll(struct target *target)
+{
+	enum target_state old_state = target->state;
+	int ret;
+
+	ret = xtensa_poll(target);
+
+	if (old_state != TARGET_HALTED && target->state == TARGET_HALTED) {
+		/*Call any event callbacks that are applicable */
+		if (old_state == TARGET_DEBUG_RUNNING)
+			target_call_event_callbacks(target, TARGET_EVENT_DEBUG_HALTED);
+		else {
+			xtensa_fileio_detect_proc(target);
+			target_call_event_callbacks(target, TARGET_EVENT_HALTED);
+		}
+	}
+
+	return ret;
+}
+
+/* 
+ * TODO: Remove if possible
+ *
+The TDI pin is also used as a flash Vcc bootstrap pin. If we reset the CPU externally, the last state of the TDI pin can
+allow the power to an 1.8V flash chip to be raised to 3.3V, or the other way around. 
+*/
+void xtensa_chip_queue_tdi_idle(struct target *target)
+{
+	static uint8_t value = 0;
+	uint8_t t[4] = { 0, 0, 0, 0 };
+
+#if 0
+	struct xtensa *xtensa = target_to_xtensa(target);
+	struct xtensa_chip_common *xtensa_chip = xtensa->xtensa_chip;
+	if (xtensa_chip->flash_bootstrap == FBS_TMSLOW) {
+		/*Make sure tdi is 0 at the exit of queue execution */
+		value = 0;
+	} else if (xtensa_chip->flash_bootstrap == FBS_TMSHIGH) {
+		/*Make sure tdi is 1 at the exit of queue execution */
+		value = 1;
+	} else
+		return;
+#endif
+
+	/*Scan out 1 bit, do not move from IRPAUSE after we're done. */
+	buf_set_u32(t, 0, 1, value);
+	jtag_add_plain_ir_scan(1, t, NULL, TAP_IRPAUSE);
+}
+
+static int xtensa_chip_virt2phys(struct target *target,
+	target_addr_t virtual, target_addr_t *physical)
+{
+	*physical = virtual;
```

Comment:
From your review comment : )) 

is pointer guaranteed non-null or should it be checked here?

Reply:
=)